File: go/zipLinkedList/zipLinkedList.go
package main

import "fmt"

type LinkedList struct {
	Value int
	Next *LinkedList
}


func ZipLinkedList(head *LinkedList) *LinkedList {
	var count int
	curr := head

	for curr != nil {
		count++
		curr = curr.Next
	}

	if count == 1 {
		return head
	}

	var startNorm, endNorm *LinkedList 
	var mid int

	if count % 2 == 0 {
		mid = count / 2
	}else {
		mid = (count / 2) + 1
	}

	for mid > 0 {
		if startNorm == nil {
			startNorm = head
			endNorm = startNorm
			mid--
		}else {
			endNorm = endNorm.Next
			mid--
		}
	}

	var startZip, endZip *LinkedList = nil, endNorm.Next

	for endZip != nil {
		point := endZip.Next
		endZip.Next = startZip
		startZip = endZip
		endZip = point
	}
	endNorm.Next = nil

	var zippedList, zippedHead *LinkedList

	for startNorm != nil {
		if zippedList ==  nil {
		zippedList = startNorm
		zippedHead = zippedList
		startNorm = startNorm.Next
		zippedList.Next = startZip
		startZip = startZip.Next
		zippedList = zippedList.Next
	}else {
		zippedList.Next = startNorm
		startNorm = startNorm.Next
		zippedList = zippedList.Next
		if startZip == nil {
			zippedList.Next = startNorm
			break
		}
		zippedList.Next = startZip
		startZip = startZip.Next
		zippedList = zippedList.Next

		}
	}
	
	return zippedHead
}


func main() {
	list := []int{1}
	head := &LinkedList{}
	curr := head
	for i, val := range list {
		curr.Value = val
		if i == len(list) - 1 {continue}
		curr.Next = &LinkedList{}
		curr = curr.Next
	}

	rearranged := ZipLinkedList(head)
	for rearranged != nil {
		fmt.Printf("%v -> ", rearranged.Value)
		rearranged = rearranged.Next
	}
	fmt.Println()
}

File: python/zipLinkedList.py
class LinkedList:
    def __init__(self, value):
        self.value = value
        self.next = None
        

def zipLinkedList(head):
    count = 0
    
    curr = head
    
    while curr != None:
        count += 1
        curr = curr.next
    
    if count == 1:
        return head
    
    startNorm, endNorm = None, None
    
    mid = 0
    if count % 2 == 0:
        mid = count / 2
    else:
        mid = (count // 2 ) + 1
    
    while mid > 0:
        if startNorm == None:
            startNorm = head
            endNorm = startNorm
            mid -= 1
        else:
            endNorm = endNorm.next
            mid -= 1
    
    startZip, endZip = None, endNorm.next
    
    while endZip != None:
        point = endZip.next
        endZip.next = startZip
        startZip = endZip
        endZip = point
        
    endNorm.next = None
    
    zippedList, zippedHead = None, None
    
    while startNorm != None:
        if zippedList == None:
            zippedList = startNorm
            zippedHead = zippedList
            startNorm = startNorm.next
            zippedList.next = startZip
            startZip = startZip.next
            zippedList = zippedList.next
    
        else:
            zippedList.next = startNorm
            startNorm = startNorm.next
            zippedList = zippedList.next
            if startZip == None:
                break
            zippedList.next = startZip
            startZip = startZip.next
            zippedList = zippedList.next
      
    return zippedHead


head = LinkedList(0)
curr = head
array = [3, 0, 5, 2, 1, 4]

for val in range(len(array)):
    curr.value = array[val]
    if val == len(array) - 1:
        break
    curr.next = LinkedList(0)
    curr = curr.next
    
arrange = zipLinkedList(head)

while arrange != None:
    print(arrange.value, end="->")
    arrange = arrange.next
print()